core/src/main/java/org/apache/calcite/sql/validate/SqlValidatorImpl.java

```diff
@@ -7204,6 +7227,146 @@ protected SqlNode expandDynamicStar(SqlIdentifier id, SqlIdentifier fqId) {
     }
   }
 
+  /**
+   * Helper function called in ExtendedAliasExpander (used to expand identifiers in the Select list,
+   * Where, and On clauses) and ExtendedExpander (used by having/group by clauses).
+   *
+   * Finds expressions that alias to "name", in the select list.
+   *
+   * @param name The alias to search for in the select list
+   * @param nameMatcher The nameMatcher used to check if the aliases match.
+   * @param select The select to search.
+   * @param maxIdx The maximum idx into the select list to search.
+   * @return A nested pair containing the number of occurrences of the alias, one of the expressions
+   * to which the alias evaluates (if there are multiple), and the index into the select list that
+   * said expression occurs at. The information is arranged (numOccurrences, (expr, expr_idx))
+   */
+  protected static Pair<Integer, Pair<@Nullable SqlNode, Integer>> findAliasesInSelect(
+      String name, SqlNameMatcher nameMatcher, SqlSelect select, Integer maxIdx) {
+    SqlNode expr = null;
+    int numOccurrencesOfAlias = 0;
+    int idxOfOccurrence = 0;
+
+    for (int i = 0; i < maxIdx; i++) {
+      SqlNode s = select.getSelectList().get(i);
+      final String alias = SqlValidatorUtil.getAlias(s, -1);
+      if (alias != null && nameMatcher.matches(alias, name)) {
+        expr = s;
+        idxOfOccurrence = i;
+        numOccurrencesOfAlias++;
+      }
+    }
+    return new Pair<>(numOccurrencesOfAlias, new Pair<>(expr, idxOfOccurrence));
+  }
+
+  /**
+   * Shuttle which walks over an expression replacing usage of alias with underlying expression.
+   */
+  static class ExtendedAliasExpander extends Expander {
+    SqlSelect aliasSourceSelect;
+
+    // When recursively checking for aliases, we often encounter situations where we need to check
+    // a subset of the select items for an alias. IE:
+    //    SELECT A, B, C from ...
+    // If we're trying to find aliases for "B" we should only check every select statement that
```

Comment:
What happens is the alias is seen multiple times? If its ambiguous it should raise an exception, but I think mimicking your example below:

```SQL
SELECT C as x, x, A + x from
```

this should be okay.

Reply:
This is probably an edge case though so if we throw an exception even in this case that would be fine with me.

Reply:
The behavior (both in SF and locally) is that a failure, with the message that x is an ambiguous column name. I'll add a test for it.